test/infrastructure/docker/internal/controllers/devmachine_controller.go

```diff
@@ -0,0 +1,257 @@
+	// Specifically, it will add KubeadmControlPlane, MachineSet and MachineDeployment.
+	ctx, log, err := clog.AddOwners(ctx, r.Client, devMachine)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// Fetch the Machine.
+	machine, err := util.GetOwnerMachine(ctx, r.Client, devMachine.ObjectMeta)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	if machine == nil {
+		log.Info("Waiting for Machine Controller to set OwnerRef on DevMachine")
+		return ctrl.Result{}, nil
+	}
+
+	log = log.WithValues("Machine", klog.KObj(machine))
+	ctx = ctrl.LoggerInto(ctx, log)
+
+	// Fetch the Cluster.
+	cluster, err := util.GetClusterFromMetadata(ctx, r.Client, machine.ObjectMeta)
+	if err != nil {
+		log.Info("DevMachine owner Machine is missing cluster label or cluster does not exist")
+		return ctrl.Result{}, err
+	}
+	if cluster == nil {
+		log.Info(fmt.Sprintf("Please associate this machine with a cluster using the label %s: <name of cluster>", clusterv1.ClusterNameLabel))
+		return ctrl.Result{}, nil
+	}
+
+	log = log.WithValues("Cluster", klog.KObj(cluster))
+	ctx = ctrl.LoggerInto(ctx, log)
+
+	if isPaused, conditionChanged, err := paused.EnsurePausedCondition(ctx, r.Client, cluster, devMachine); err != nil || isPaused || conditionChanged {
+		return ctrl.Result{}, err
+	}
+
+	if cluster.Spec.InfrastructureRef == nil {
+		log.Info("Cluster infrastructureRef is not available yet")
+		return ctrl.Result{}, nil
+	}
+
+	// Fetch the DevCluster.
+	devCluster := &infrav1.DevCluster{}
+	devClusterName := client.ObjectKey{
+		Namespace: devMachine.Namespace,
+		Name:      cluster.Spec.InfrastructureRef.Name,
+	}
+	if err := r.Client.Get(ctx, devClusterName, devCluster); err != nil {
+		log.Info("DockerCluster is not available yet")
+		return ctrl.Result{}, nil
+	}
+
+	backendReconciler := r.backendReconcilerFactory(ctx, devMachine)
+
+	// Initialize the patch helper
+	patchHelper, err := patch.NewHelper(devMachine, r)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	// Always attempt to Patch the DevMachine object and status after each reconciliation.
+	defer func() {
+		if err := backendReconciler.PatchDevMachine(ctx, patchHelper, devMachine, util.IsControlPlaneMachine(machine)); err != nil {
+			rerr = kerrors.NewAggregate([]error{rerr, err})
+		}
+	}()
+
+	// Handle deleted machines
+	if !devMachine.ObjectMeta.DeletionTimestamp.IsZero() {
+		return backendReconciler.ReconcileDelete(ctx, cluster, devCluster, machine, devMachine)
+	}
+
+	// Handle non-deleted machines
+	res, err := backendReconciler.ReconcileNormal(ctx, cluster, devCluster, machine, devMachine)
+	// Requeue if the reconcile failed because the ClusterCacheTracker was locked for
+	// the current cluster because of concurrent access.
+	if errors.Is(err, clustercache.ErrClusterNotConnected) {
+		log.V(5).Info("Requeuing because connection to the workload cluster is down")
+		return ctrl.Result{RequeueAfter: time.Minute}, nil
+	}
```

Comment:
This was removed in other controllers. Please diff against the docker machine controller